lib/FindTests.js

```diff
@@ -0,0 +1,269 @@
+            )
+          );
+      }
+
+      // By finding the module name from the file path we can import it even if
+      // the file is full of errors. Elm will then report what’s wrong.
+      const moduleNameParts = path
+        .relative(matchingSourceDirs[0], filePath)
+        .replace(/\.elm$/, '')
+        .split(path.sep);
+      const moduleName = moduleNameParts.join('.');
+
+      if (!moduleNameParts.every(Parser.isUpperName)) {
+        return Promise.reject(
+          new Error(
+            badModuleNameError(filePath, matchingSourceDirs[0], moduleName)
+          )
+        );
+      }
+
+      return Parser.extractExposedPossiblyTests(
+        filePath,
+        // We’re reading files asynchronously in a loop here, so it makes sense
+        // to use graceful-fs to avoid “too many open files” errors.
+        gracefulFs.createReadStream
+      ).then((possiblyTests) => ({
+        moduleName,
+        possiblyTests,
+      }));
+    })
+  );
+}
+
+function missingSourceDirectoryError(filePath, isPackageProject) {
+  return `
+This file:
+
+${filePath}
+
+…matches no source directory! Imports won't work then.
+
+${
+  isPackageProject
+    ? 'Move it to tests/ or src/ in your project root.'
+    : 'Move it to tests/ in your project root, or make sure it is covered by "source-directories" in your elm.json.'
+}
+  `.trim();
+}
+
+function multipleSourceDirectoriesError(
+  filePath,
+  matchingSourceDirs,
+  testsDir
+) {
+  const note = matchingSourceDirs.includes(testsDir)
+    ? "Note: The tests/ folder counts as a source directory too (even if it isn't listed in your elm.json)!"
+    : '';
+
+  return `
+This file:
+
+${filePath}
+
+…matches more than one source directory:
+
+${matchingSourceDirs.join('\n')}
+
+Edit "source-directories" in your elm.json and try to make it so no source directory contains another source directory!
+
+${note}
+  `.trim();
+}
+
+function badModuleNameError(filePath, sourceDir, moduleName) {
+  return `
+This file:
+
+${filePath}
+
+…located in this directory:
+
+${sourceDir}
+
+…is problematic. Trying to construct a module name from the parts after the directory gives:
+
+${moduleName}
+
+…but module names need to look like for example:
+
+Main
+Http.Helpers
+
+Make sure that all parts start with an uppercase letter and don't contain any spaces or anything like that.
+  `.trim();
+}
+
+function noFilesFoundError(
+  projectRootDir /*: string */,
+  testFileGlobs /*: Array<string> */
+) /*: string */ {
+  return testFileGlobs.length === 0
+    ? `
+${noFilesFoundInTestsDir(projectRootDir)}
+
+To generate some initial tests to get things going: elm-test init
+
+Alternatively, if your project has tests in a different directory,
+try calling elm-test with a glob such as: elm-test "src/**/*Tests.elm"
+      `.trim()
+    : `
+No files found matching:
+
+${testFileGlobs.join('\n')}
+
+Are the above patterns correct? Maybe try running elm-test with no arguments?
+      `.trim();
+}
```

Comment:
From elm-test.js. The only change is the first line of the error message – it used to always be “No .elm files found in the tests/ directory.”. Now, it depends on the `noFilesFoundInTestsDir` function – see the next review comment.